[ macOS, iOS ] ASSERTION FAILED: remoteWorkerProcesses().contains(process) in WebKit::WebProcessPool::removeFromRemoteWorkerProcesses

cdumez opened the pull request:
#### ce9cd2b20a505c1e2bb9b135fc7e29db3bd1b0e9
<pre>
[ macOS, iOS ] ASSERTION FAILED: remoteWorkerProcesses().contains(process) in WebKit::WebProcessPool::removeFromRemoteWorkerProcesses
<a href="https://bugs.webkit.org/show_bug.cgi?id=256233">https://bugs.webkit.org/show_bug.cgi?id=256233</a>
rdar://104915398

Reviewed by Alex Christensen.

I haven&apos;t been able to reproduce the assertion hit but I am refactoring the code to
make it unlikely the state gets out of sync between the WebProcessPool and the
WebProcessProxy.

The WebProcessPool used to add the WebProcessProxy to the remoteWorkerProcesses()
HashMap. The removal used to happen both from the WebProcessPool and the
WebProcessProxy, depending on the reason.

The new logic is simplified as the WebProcessProxy is now solely in charge of adding
and removing itself from the map. This should make sure that WebProcessProxy cannot
be running remote workers and yet not be in the WebProcessPool map (like the assertion
suggests).

* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::establishRemoteWorkerContextConnectionToNetworkProcess):
(WebKit::WebProcessPool::addRemoteWorkerProcess):
(WebKit::WebProcessPool::removeRemoteWorkerProcess):
(WebKit::WebProcessPool::disconnectProcess):
(WebKit::WebProcessPool::removeFromRemoteWorkerProcesses): Deleted.
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::disableRemoteWorkers):
(WebKit::WebProcessProxy::enableRemoteWorkers):

Canonical link: <a href="https://commits.webkit.org/263693@main">https://commits.webkit.org/263693@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ef1f1341243fc322d783f2f0f4650e6e00fe747b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5436 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5576 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5768 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6978 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5464 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5805 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5554 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/6371 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5538 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5601 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4870 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7008 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3079 "Passed tests") | | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/12024 "142 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4945 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4955 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6669 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5391 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4425 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4851 "Built successfully") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/1305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8952 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5214 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -1006,12 +1013,15 @@ void WebProcessPool::disconnectProcess(WebProcessProxy& process)
-    if (process.isRunningWorkers())
-        removeFromRemoteWorkerProcesses(process);
+    if (process.isRunningWorkers()) {
+        process.disableRemoteWorkers(RemoteWorkerType::ServiceWorker);
```
These two functions are called together twice.  Maybe make it take an OptionSet?

cdumez replied:
Good idea. Will do.

cdumez committed:
[ macOS, iOS ] ASSERTION FAILED: remoteWorkerProcesses().contains(process) in WebKit::WebProcessPool::removeFromRemoteWorkerProcesses
https://bugs.webkit.org/show_bug.cgi?id=256233
rdar://104915398

Reviewed by Alex Christensen.

I haven't been able to reproduce the assertion hit but I am refactoring the code to
make it unlikely the state gets out of sync between the WebProcessPool and the
WebProcessProxy.

The WebProcessPool used to add the WebProcessProxy to the remoteWorkerProcesses()
HashMap. The removal used to happen both from the WebProcessPool and the
WebProcessProxy, depending on the reason.

The new logic is simplified as the WebProcessProxy is now solely in charge of adding
and removing itself from the map. This should make sure that WebProcessProxy cannot
be running remote workers and yet not be in the WebProcessPool map (like the assertion
suggests).

* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::establishRemoteWorkerContextConnectionToNetworkProcess):
(WebKit::WebProcessPool::addRemoteWorkerProcess):
(WebKit::WebProcessPool::removeRemoteWorkerProcess):
(WebKit::WebProcessPool::disconnectProcess):
(WebKit::WebProcessPool::removeFromRemoteWorkerProcesses): Deleted.
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::disableRemoteWorkers):
(WebKit::WebProcessProxy::enableRemoteWorkers):

Canonical link: https://commits.webkit.org/263693@main
